### ISSUE TITLE
Panicking in `cx.spawn` task should not bring down the entire app

### ISSUE THREAD

**Issue by jkelleyrtp** [DioxusLabs/dioxus#632]
## Specific Demand

In react, if you throw an exception, it'll get caught by the react machinery as an error at the nearest error boundary.

WASM does not have the ability to catch panics properly, but non-wasm targets do.

We should catch panics in tasks and suspense and bubble them to the nearest error boundary if possible.

This is pretty straightforward by simply wrapping the scheduler poll logic with a panic handler.

## Implement Suggestion

This comes in two parts
- [ ] Add panic handling to rust for wasm targets
- [ ] Wrap the scheduler logic with a panic handler

**Comment by jkelleyrtp** [comment]
From evan in the discord:
---
Have you run into unwind safety before?
Both tasks and event handling use interior mutability and can run user code which contains interior mutability. That causes trying to catch unwinds to give an error about unwind safety 
We can assert that it is unwind safe, but I'm not sure it is
This isn't a safety issue, but it can cause logic errors

Types such as &mut T and &RefCell<T> are examples which are not unwind safe. The general idea is that any mutable state which can be shared across catch_unwind is not unwind safe by default. This is because it is very easy to witness a broken invariant outside of catch_unwind as the data is simply accessed as usual.
---

Seems like it could be possible, but requires us to think about the invariants we're upholding in specific chunks of the codebase.

Do we wrap .poll() or do we wrap the whole method? Where do we need to focus on maintaining invariants?

Ideally this is just a debug_assertions feature so invariants aren't terribly important - double panics are okay in debug mode.

**Comment by chungwong** [comment]
Should this issue be considered adding to the 0.6 release/milestone?
Given that panics could happens in other deps and potentially uncontrolled, it is very easy to crash the entire app.

**Comment by ealmloff** [comment]
> Should this issue be considered adding to the 0.6 release/milestone?
> Given that panics could happens in other deps and potentially uncontrolled, it is very easy to crash the entire app.

Libraries should return `Result`s for actions that can fail. This should help avoid issues with unexpected panics on desktop, but it shouldn't be something you rely on for known panics. This isn't possible to implement consistently across all platforms because rust compiled to wasm doesn't support unwinding. It should act more as a safe guard if all other error handling fails

**Comment by chungwong** [comment]
Maybe I should talk about the issue with concrete examples rather than hypothetical.
There are lots of cases where a function can panics.
1. [Vec::with_capacity](https://doc.rust-lang.org/std/vec/struct.Vec.html#panics)
2. [Vec::windows](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.windows)

Of course, those panics are known and arguably can be handled better. Let's actually get to the conrete part and focus on https://dioxuslabs.com which is an excellent example to demostrate the concern I have.

https://github.com/DioxusLabs/docsite/issues/288
https://github.com/DioxusLabs/docsite/issues/218
https://discord.com/channels/899851952891002890/1275886621597634670

Because of the panics, none of the WASM stuff worked, coutner example didn't work, search didn't work, the nav hamburger didn't work. So panics are actually unpredictable  and the results are not desirable in practice.

**Comment by jkelleyrtp** [comment]
because wasm doesn't support unwinding, we can't support this in any reasonable time frame.

**Comment by chungwong** [comment]
> because wasm doesn't support unwinding, we can't support this in any reasonable time frame.


Does it mean https://github.com/WebAssembly/spec/blob/wasm-3.0/proposals%2Fexception-handling%2FExceptions.md wouldn't help this issue!

**Comment by jkelleyrtp** [comment]
> https://github.com/WebAssembly/spec/blob/wasm-3.0/proposals%2Fexception-handling%2FExceptions.md

Exceptions in wasm3 would help and let us properly handle panics.

I saw your message re: the page bricking on panic, and found a place where we were holding a refcell over user callback, and then fixed it.

https://github.com/DioxusLabs/dioxus/pull/5489

**Comment by chungwong** [comment]
I just did a comparison between Leptos and Dioxus

In Leptos, 
```rs
#[component]
fn App() -> impl IntoView {
    let (count, set_count) = signal(0);

    view! {
        <button on:click=move |_| set_count.update(|c| *c += 1)>
            {move || count.get()}
        </button>
        <button on:click=|_| panic!("this one panics")>"Uh oh"</button>
    }
}
```
The button caused a panic but the whole app was still usable, the other button was clickable after the panic happened.

This is not the case in Dioxus, any panic will just kill the whole app and the users need to refresh the page to start over.
```rsx
#[component]
fn App() -> Element {
    let mut count = use_signal(|| 0);

    rsx! {
        h1 { "High-Five counter: {count}" }
        button { onclick: move |_| count += 1, "Up high!" }
        button { onclick: move |_| count -= 1, "Down low!" }
        button {
            onclick: move |_| -> () {
                panic!("dddd");
            },
            "panick"
        }
    }
}
```
Can I confirm the fix actually make the Leptos and Dioxus behave the same, i.e. not rendering the whole app useless?

**Comment by jkelleyrtp** [comment]
> I just did a comparison between Leptos and Dioxus
> 
> In Leptos, 
> ```rs
> #[component]
> fn App() -> impl IntoView {
>     let (count, set_count) = signal(0);
> 
>     view! {
>         <button on:click=move |_| set_count.update(|c| *c += 1)>
>             {move || count.get()}
>         </button>
>         <button on:click=|_| panic!("this one panics")>"Uh oh"</button>
>     }
> }
> ```
> The button caused a panic but the whole app was still usable, the other button was clickable after the panic happened.
> 
> This is not the case in Dioxus, any panic will just kill the whole app and the users need to refresh the page to start over.
> ```rsx
> #[component]
> fn App() -> Element {
>     let mut count = use_signal(|| 0);
> 
>     rsx! {
>         h1 { "High-Five counter: {count}" }
>         button { onclick: move |_| count += 1, "Up high!" }
>         button { onclick: move |_| count -= 1, "Down low!" }
>         button {
>             onclick: move |_| -> () {
>                 panic!("dddd");
>             },
>             "panick"
>         }
>     }
> }
> ```
> Can I confirm the fix actually make the Leptos and Dioxus behave the same, i.e. not rendering the whole app useless?

Yes I used that very example to test out this panic situation. Dioxus still chugs along even after the panic after this fix.
